src/transform/src/fusion/flatmap_to_map.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Turns `FlatMap` into `Map` if only one row is produced by flatmap.
+//!
+
+use crate::TransformArgs;
+use mz_expr::{MirRelationExpr, TableFunc};
+
+/// Turns `FlatMap` into `Map` if only one row is produced by flatmap.
+#[derive(Debug)]
+pub struct FlatMapToMap;
+
+impl crate::Transform for FlatMapToMap {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.try_visit_mut_post(&mut |e| Ok(self.action(e)))
+    }
+}
+
+impl FlatMapToMap {
+    fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::FlatMap { func, exprs, .. } = relation {
+            if let TableFunc::Wrap { width, .. } = func {
+                if *width == exprs.len() {
```

Comment:
Alternatively, you can keep the `*width == exprs.len()` guard unchanged, but then change

1. the implementation of `TableFunc::eval` for `Wrap` and 
2. the delegated `wrap` function that implements the evaluation logic

so that it's return type of the latter is `Result<Box<dyn Iterator<Item = (Row, Diff)> + 'a>, EvalError>` and an `Result::Err` is returned if `width` doesn't divide `datums` without remainder.

Reply:
Ok, that makes sense. I have changed it to `>=`